llvm/lib/Target/X86/MCTargetDesc/X86MCCodeEmitter.cpp

```diff
@@ -139,7 +147,20 @@ class X86OpcodePrefixHelper {
   }
 
   void setR(unsigned Encoding) { R = Encoding >> 3 & 1; }
-  void setR2(unsigned Encoding) { EVEX_R2 = Encoding >> 4 & 1; }
+  void setR2(unsigned Encoding) {
+    R2 = Encoding >> 4 & 1;
+    assert((!R2 || (Kind <= REX2 || Kind == EVEX)) && "invalid setting");
```

Comment:
`Kind <= REX2` includes `Kind == None` which doesn't seem correct.
The same for below.

Reply:
It does include `Kind == None` and it's expected. `determineOptimalKind()` is called at line 1354 to get the optimal prefix and any `Kind <= REX2` may be promoted to `REX2`.